### ISSUE TITLE
cpuid_get_total_cpus returns wrong CPU count

### ISSUE THREAD

**Issue by S74nk0** [anrieff/libcpuid#75]
Hello!
I am using libcpuid for a cross platform project. Everything works great except the "**int cpuid_get_total_cpus (  void  )**" function it returns 8 instead of 1 on Windows 10 64bit (It has one physical CPU with 8 virtual cores). From the documentation and function name one would expect to get the count of physical CPUs but it seems to be returning all available virtual cores. Is this a BUG or have I misunderstood the documentation? If the library does not have a way to get physical CPU count please let me know, so I will fork and implement this feature (Windows and Linux 64bit), or if it is a BUG will fix it.
Thanks again for the great project.

**Comment by anrieff** [comment]
Hi!
Thank you for your report. The function you named is badly named in the sense that it really returns the number of virtual cores. I'll update the documentation to make it clear what it does.

On your question, what you probably would need is to call `cpu_identify()` and then look into the following items of the `cpu_id_t` structure:

	/** Number of CPU cores on the current processor */
	int32_t num_cores;
	
	/**
	 * Number of logical processors on the current processor.
	 * Could be more than the number of physical cores,
	 * e.g. when the processor has HyperThreading.
	 */
	int32_t num_logical_cpus;
	
	/**
	 * The total number of logical processors.
	 *
	 * This is num_logical_cpus * {total physical processors in the system}
	 *
	 * If you're writing a multithreaded program and you want to run it on
	 * all CPUs, this is the number of threads you need.
	 */
	int32_t total_logical_cpus;

Imagine a system with 2 CPU sockets, each hosting a processor with 4 cores and each core has hyper-threading.

- `cpu_id_t::num_cores`  would be 4, since on the CPU you are querying, you have only 4 cores;

- `cpu_id_t::num_logical_cpus` would be 8, since that's how many threads you have on the current processor, considering the HT;

- `cpu_id_t::total_logical_cpus` would be 16, since that's how many threads you have in the whole system.

`cpuid_get_total_cpus()` actually returns `total_logical_cpus`.

You can compute the number of physical CPUs indirectly, using

`physical_cpus = total_logical_cpus / num_logical_cpus`.

One bit of VM-specific pitfall you might have issues with is that `total_logical_cpus`, as well as the `cpuid_get_total_cpus()` function are implemented by common code and it uses OS routines to always get the real available hardware threads your program can run on. You might have a multicore CPU, but if you created a VM and limited the number of threads to just 1, inside the VM you'd get `cpuid_get_total_cpus() == cpu_id_t::total_logical_cpus == 1`. So under a VM, the values you will get might be mangled. See, for example, issue #50.

In a nutshell, if you're on a VM, you are on your own. Beware of the dragons :)

**Comment by S74nk0** [comment]
Hi @anrieff. Thanks for the quick feedback :). Your answer is clean and concise, I understand now how to use the library correctly and also thanks for pointing out potential issues for VMs :). I will close this issue. Thanks again!
